src/rgw/rgw_auth_s3.cc

```diff
@@ -239,199 +240,819 @@ bool rgw_create_s3_canonical_header(const req_info& info,
+      token_env = "CONTENT_LENGTH";
+    }
+    if (token_env == "HTTP_CONTENT_TYPE") {
+      token_env = "CONTENT_TYPE";
+    }
+    const char *t = info.env->get(token_env.c_str());
+    if (!t) {
+      dout(10) << "warning env var not available" << dendl;
+      continue;
+    }
+    if (token_env == "HTTP_CONTENT_MD5") {
+      for (const char *p = t; *p; p++) {
+	if (!is_base64_for_content_md5(*p)) {
+	  dout(0) << "NOTICE: bad content-md5 provided (not base64), aborting request p=" << *p << " " << (int)*p << dendl;
+	  return boost::none;
+	}
+      }
+    }
+    string token_value = string(t);
+    if (force_boto2_compat && using_qs && (token == "host")) {
+      if (!secure_port.empty()) {
+	if (secure_port != "443")
+	  token_value = token_value + ":" + secure_port;
+      } else if (!port.empty()) {
+	if (port != "80")
+	  token_value = token_value + ":" + port;
+      }
+    }
+    canonical_hdrs_map[token] = rgw_trim_whitespace(token_value);
+  }
 
-  rgw_assemble_s3_v4_canonical_request(s->info.method, canonical_uri.c_str(),
-      canonical_qs.c_str(), canonical_hdrs.c_str(), signed_hdrs.c_str(),
-      request_payload_hash.c_str(), canonical_req);
+  std::string canonical_hdrs;
+  for (map<string, string>::iterator it = canonical_hdrs_map.begin();
+      it != canonical_hdrs_map.end(); ++it) {
+    canonical_hdrs.append(it->first + ":" + it->second + "\n");
+  }
 
-  rgw_hash_s3_string_sha256(canonical_req.c_str(), canonical_req.size(), canonical_req_hash);
+  return canonical_hdrs;
+}
 
-  ldout(s->cct, 10) << "canonical request = " << canonical_req << dendl;
-  ldout(s->cct, 10) << "canonical request hash = " << canonical_req_hash << dendl;
+/*
+ * create canonical request for signature version 4
+ *
+ * http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+ */
+sha256_digest_t
+get_v4_canon_req_hash(CephContext* cct,
+                      const boost::string_ref& http_verb,
+                      const std::string& canonical_uri,
+                      const std::string& canonical_qs,
+                      const std::string& canonical_hdrs,
+                      const std::string& signed_hdrs,
+                      const boost::string_ref& request_payload_hash)
+{
+  ldout(cct, 10) << "payload request hash = " << request_payload_hash << dendl;
+
+  const auto canonical_req = std::string()
+    .append(http_verb.data(), http_verb.length())
+    .append("\n")
+    .append(canonical_uri)
+    .append("\n")
+    .append(canonical_qs)
+    .append("\n")
+    .append(canonical_hdrs)
+    .append("\n")
+    .append(signed_hdrs)
+    .append("\n")
+    .append(request_payload_hash.data(), request_payload_hash.length());
+
+  const auto canonical_req_hash = calc_hash_sha256(canonical_req);
+
+  ldout(cct, 10) << "canonical request = " << canonical_req << dendl;
+  ldout(cct, 10) << "canonical request hash = "
+                 << buf_to_hex(canonical_req_hash).data() << dendl;
```

Comment:
> this doesn't look safe - `data()` will return a pointer to memory that's destructed before it's passed as an argument to `operator<<()`

`buf_to_hex` constructs a temporary object. I believe its lifetime ends [*at the end of the full-expression*](http://stackoverflow.com/a/2506800). Let's verify that with the following snippet:

```cpp
#include <iostream>

class PrintingDtor {
public:
  const char* data() const {
    return "PrintingDtor: data() called";
  }

  ~PrintingDtor() {
    std::cout << "PrintingDtor: dtor called" << std::endl;
  }
};

static PrintingDtor lf_time() {
  return PrintingDtor();
}

int main (void) {
  std::cout << "+++ OBJ LIFETIME TEST -- start" << std::endl;
  std::cout << "+++ OBJ LIFETIME TEST: "
                 << lf_time().data() << "; print end" << std::endl;
  std::cout << "+++ OBJ LIFETIME TEST -- end" << std::endl;
}
```
```
$ /tmp/templf 
+++ OBJ LIFETIME TEST -- start
+++ OBJ LIFETIME TEST: PrintingDtor: data() called; print end
PrintingDtor: dtor called
+++ OBJ LIFETIME TEST -- end
```
The binary has been produced by GCC 4.8.4 on AMD64.